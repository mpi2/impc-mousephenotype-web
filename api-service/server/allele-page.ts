import { fetchAPIFromServer, fetchURL } from "@/api-service";
import { AlleleSummary } from "@/models";
const KUBERNETES_NAMESPACE = process.env.KUBERNETES_NAMESPACE ?? "default";
const NODE_ENV = process.env.NODE_ENV;

export async function fetchAlleleSummary(
  mgiGeneAccessionId: string,
  alleleName: string,
): Promise<AlleleSummary> {
  const endpointURL = `http://impc-allele-service.${KUBERNETES_NAMESPACE}:8080/v1/alleles?mgiGeneAccessionId=${mgiGeneAccessionId}&alleleName=${alleleName}`;
  return await (NODE_ENV === "development"
    ? fetchAPIFromServer<AlleleSummary>(
        `/api/v1/alleles/${mgiGeneAccessionId}/${alleleName}`,
      )
    : fetchURL<AlleleSummary>(endpointURL));
}
