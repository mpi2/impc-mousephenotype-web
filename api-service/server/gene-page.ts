import { GeneOrder, GenePhenotypeHits, GeneSummary } from "@/models/gene";
import { fetchAPIFromServer, fetchURL } from "@/api-service";
import { processGeneOrderResponse } from "@/hooks/gene-order.query";
import { processGenePhenotypeHitsResponse } from "@/hooks/significant-phenotypes.query";

const KUBERNETES_NAMESPACE = process.env.KUBERNETES_NAMESPACE ?? "default";
const NODE_ENV = process.env.NODE_ENV;

export async function fetchGeneSummary(
  mgiGeneAccessionId: string,
): Promise<GeneSummary> {
  const endpointURL = `http://impc-summary-service.${KUBERNETES_NAMESPACE}:8080/v1/gene_summary?mgiGeneAccessionId=${mgiGeneAccessionId}`;
  return await (NODE_ENV === "development"
    ? fetchAPIFromServer<GeneSummary>(
        `/api/v1/genes/${mgiGeneAccessionId}/summary`,
      )
    : fetchURL<GeneSummary>(endpointURL));
}

export async function fetchGenePhenotypeHits(
  mgiGeneAccessionId: string,
): Promise<Array<GenePhenotypeHits>> {
  const endpointURL = `http://impc-phenotype-hits-service.${KUBERNETES_NAMESPACE}:8080/v1/phenotypehits?mgiGeneAccessionId=${mgiGeneAccessionId}`;
  const data = await (NODE_ENV === "development"
    ? fetchAPIFromServer<Array<GenePhenotypeHits>>(
        `/api/v1/genes/${mgiGeneAccessionId}/phenotype-hits`,
      )
    : fetchURL<Array<GenePhenotypeHits>>(endpointURL));
  return processGenePhenotypeHitsResponse(data);
}

export async function fetchGeneOrderData(
  mgiGeneAccessionId: string,
): Promise<Array<GeneOrder>> {
  const endpointURL = `http://impc-order-service.${KUBERNETES_NAMESPACE}:8080/v1/orders?mgiGeneAccessionId=${mgiGeneAccessionId}`;
  const data = await (NODE_ENV === "development"
    ? fetchAPIFromServer<Array<GeneOrder>>(
        `/api/v1/genes/${mgiGeneAccessionId}/order`,
      )
    : fetchURL<Array<GeneOrder>>(endpointURL));
  return processGeneOrderResponse(data);
}
