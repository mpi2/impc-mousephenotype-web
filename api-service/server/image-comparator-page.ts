import { GeneImageCollection } from "@/models/gene";
import { fetchAPIFromServer, fetchURL } from "@/api-service";

const KUBERNETES_NAMESPACE = process.env.KUBERNETES_NAMESPACE ?? "default";
const NODE_ENV = process.env.NODE_ENV;

export async function fetchMutantImages(
  mgiGeneAccessionId: string,
  parameterStableId: string,
): Promise<Array<GeneImageCollection>> {
  const endpointURL = `http://impc-imaging-service.${KUBERNETES_NAMESPACE}:8080/v1/images/findByMgiAndStableId?mgiGeneAccessionId=${mgiGeneAccessionId}&parameterStableId=${parameterStableId}`;
  try {
    return await (NODE_ENV === "development"
      ? fetchAPIFromServer<Array<GeneImageCollection>>(
          `/api/v1/images/find_by_mgi_and_stable_id?mgiGeneAccessionId=${mgiGeneAccessionId}&parameterStableId=${parameterStableId}`,
        )
      : fetchURL<Array<GeneImageCollection>>(endpointURL));
  } catch (error) {
    return [];
  }
}

export async function fetchControlImages(
  parameterStableId: string,
): Promise<Array<GeneImageCollection>> {
  const endpointURL = `http://impc-imaging-service.${KUBERNETES_NAMESPACE}:8080/v1/images/findByStableIdAndSampleId?biologicalSampleGroup=control&parameterStableId=${parameterStableId}`;
  try {
    return await (NODE_ENV === "development"
      ? fetchAPIFromServer<Array<GeneImageCollection>>(
          `/api/v1/images/find_by_stable_id_and_sample_id?biologicalSampleGroup=control&parameterStableId=${parameterStableId}`,
        )
      : fetchURL<Array<GeneImageCollection>>(endpointURL));
  } catch (error) {
    return [];
  }
}
