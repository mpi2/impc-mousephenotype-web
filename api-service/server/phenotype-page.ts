import { fetchAPIFromServer, fetchURL } from "@/api-service";
import { PhenotypeSummary } from "@/models/phenotype";

const KUBERNETES_NAMESPACE = process.env.KUBERNETES_NAMESPACE ?? "default";
const NODE_ENV = process.env.NODE_ENV;

export async function fetchPhenotypeSummary(
  phenotypeId: string,
): Promise<PhenotypeSummary> {
  const endpointURL = `http://impc-summary-service.${KUBERNETES_NAMESPACE}:8080/v1/phenotype_summary?phenotypeId=${phenotypeId}`;
  return await (NODE_ENV === "development"
    ? fetchAPIFromServer<PhenotypeSummary>(
        `/api/v1/phenotypes/${phenotypeId}/summary`,
      )
    : fetchURL<PhenotypeSummary>(endpointURL));
}
