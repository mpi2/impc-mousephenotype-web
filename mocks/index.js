export async function initMocks() {
  if (typeof window === "undefined") {
    const { server } = await import("./server");
    server.listen();
  } else {
    const { worker } = await import("./browser");
    return worker.start({
      serviceWorker: {
        url: "/data/mockServiceWorker.js",
        options: {
          scope: "/",
        }
      }
    });
  }
}

