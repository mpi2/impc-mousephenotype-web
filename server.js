const { createServer } = require('http');
const { parse } = require('url');
const next = require('next');

const dev = process.env.NODE_ENV !== 'production';
const hostname = 'localhost';
const port = 3000;
const app = next({ dev, hostname, port });
const handle = app.getRequestHandler();

app.prepare()
.then(() => {
    createServer(async (req, res) => {
        try {
            const parsedUrl = parse(req.url, true);
            const { pathname, query} = parsedUrl;
            const host = req.headers.host;
            console.log({ host, pathname });
            res.setHeader('X-IMPC-TEST', "true");
            if (host.includes('nginx.mousephenotype-dev.org')) {
                const modifiedPathName = pathname.replace("/data/", "");
                console.log({ modifiedPathName });
                await app.render(req, res, modifiedPathName, query);
                res.setHeader('X-PATH', JSON.stringify({ host, pathname, modifiedPathName }));
            } else {
                await handle(req, res, parsedUrl);
            }

        } catch (err) {
            console.error('Error occurred handling', req.url, err)
            res.statusCode = 500
            res.end('internal server error')
        }
    })
        .once('error', (err) => {
            console.error(err);
            process.exit(1);
        })
        .listen(port, () => {
            console.log(`Ready on http://${hostname}:${port}`);
        })
});