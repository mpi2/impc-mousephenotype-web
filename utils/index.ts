import _ from "lodash";

export const formatBodySystems = (systems: string[] | string = []) => {
  return _.capitalize(
    (typeof systems === "string" ? systems : systems.join(", "))
      .replace(/ phenotype/g, "")
      .replace(/\//g, " / ")
  );
};

export const formatAlleleSymbol = (allele: string) => {
  return allele.slice(0, allele.length - 1).split("<");
};
