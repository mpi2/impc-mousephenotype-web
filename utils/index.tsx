import React from "react";
import _ from "lodash";

export const formatBodySystems = (systems: string[] | string = []) => {
  return _.capitalize(
    (typeof systems === "string" ? systems : systems.join(", "))
      .replace(/ phenotype/g, "")
      .replace(/\//g, " / ")
  );
};

export const formatAlleleSymbol = (allele: string) => {
  return allele.slice(0, allele.length - 1).split("<");
};

export const formatPValue = (pValue: number) => {
  const pValueArray = Number.parseFloat(String(pValue))
    .toExponential(2)
    .split("e");
  return (
    <>
      {pValueArray[0]}x10<sup>{pValueArray[1].replace("+", "")}</sup>
    </>
  );
};

export const formatESCellName = (src: string) => {
  const [name, superscript, end] = src.split(/<|>/);
  return (
    <>
      {name}
      <sup>{superscript}</sup>
      {end}
    </>
  );
};

const fetchCache = {};
export { fetchCache };

export const toSentenceCase = (camelCase) => {
  if (camelCase) {
    const result = camelCase.replace(/([A-Z])/g, " $1");
    return result[0].toUpperCase() + result.substring(1).toLowerCase();
  }
  return "";
};
