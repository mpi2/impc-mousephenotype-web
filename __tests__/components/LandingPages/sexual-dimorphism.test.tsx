import { render, waitFor } from "@testing-library/react";
import SexualDimorphismLandingPage from "@/app/sexual-dimorphism/sexual-dimorphism-page";
import { server } from "../../../mocks/server";
import { rest } from "msw";
import { API_URL, createTestQueryClient } from "../../utils";
import { QueryClientProvider } from "@tanstack/react-query";

window.ResizeObserver =
  window.ResizeObserver ||
  jest.fn().mockImplementation(() => ({
    disconnect: jest.fn(),
    observe: jest.fn(),
    unobserve: jest.fn(),
  }));

describe("Sexual dimorphism landing page", () => {
  it("renders correctly", async () => {
    const client = createTestQueryClient();
    server.use(
      rest.get(`${API_URL}/api/v1/publications`, (req, res, ctx) => {
        return res(
          ctx.status(200),
          ctx.json({
            content: [],
            first: true,
            last: false,
            number: 0,
            numberOfElements: 0,
            size: 10,
            totalElements: 0,
            totalPages: 0,
          }),
        );
      }),
    );
    const { container } = render(
      <QueryClientProvider client={client}>
        <SexualDimorphismLandingPage />
      </QueryClientProvider>,
    );
    expect(container).toMatchSnapshot();
  });
});
