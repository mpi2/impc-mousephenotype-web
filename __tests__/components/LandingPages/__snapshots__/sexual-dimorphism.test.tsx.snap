// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sexual dimorphism landing page renders correctly 1`] = `
<div>
  <div
    class="banner"
  >
    <div
      class="pb-4 pt-5 container container"
    >
      <div
        class="col-12 col-md-8 ps-4 pe-4"
      >
        <div
          class="tabs"
        >
          <a
            class="tab__active"
            href="/search"
          >
            Genes
          </a>
          <a
            class="tab"
            href="/search?type=pheno"
          >
            Phenotypes
          </a>
          <a
            class="tab"
            href="https://dev.mousephenotype.org/?s="
          >
            Help, news, blog
          </a>
        </div>
        <div
          class="inputCont"
        >
          <input
            class="input"
            placeholder="Search for a gene..."
            title="main search box"
            type="text"
            value=""
          />
          <button
            aria-describedby="svg-inline--fa-title-search-icon"
            class="searchBtn"
          >
            <svg
              aria-labelledby="svg-inline--fa-title-search-icon"
              class="svg-inline--fa fa-magnifying-glass"
              data-icon="magnifying-glass"
              data-prefix="fas"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <title
                id="svg-inline--fa-title-search-icon"
              >
                Search button
              </title>
              <path
                d="M416 208c0 45.9-14.9 88.3-40 122.7L502.6 457.4c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L330.7 376c-34.4 25.2-76.8 40-122.7 40C93.1 416 0 322.9 0 208S93.1 0 208 0S416 93.1 416 208zM208 352a144 144 0 1 0 0-288 144 144 0 1 0 0 288z"
                fill="currentColor"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="page container"
    style="line-height: 2;"
  >
    <div
      class="card"
    >
      <div
        class="subheading"
      >
        <nav
          aria-label="breadcrumb"
        >
          <ol
            class="breadcrumb"
          >
            <li
              aria-current="page"
              class="breadcrumb-item active"
            >
              Home
            </li>
            <li
              aria-current="page"
              class="breadcrumb-item active"
            >
              IMPC data collections
            </li>
            <li
              aria-current="page"
              class="breadcrumb-item active"
            >
              Sexual Dimorphism Data
            </li>
          </ol>
        </nav>
      </div>
      <h1
        class="mb-4 mt-2"
      >
        <strong>
          IMPC Sexual Dimorphism Data
        </strong>
      </h1>
      <div
        class="fade alert alert-landing-page show"
        role="alert"
      >
        <div
          class="alert-heading h4"
        >
          Attention
        </div>
        <p>
          This publication page originally used the Data Release 4.2
        </p>
        <hr />
        <a
          class="link primary"
          href="https://ftp.ebi.ac.uk/pub/databases/impc/all-data-releases/release-04.2/"
        >
          Link to FTP site
        </a>
      </div>
      <div
        class="container"
      >
        <div
          class="row"
        >
          <div
            class="col-12"
          >
            <p>
              IMPC data demonstrates the effect of sex on many phenotypes, supporting the importance of including males and females in biomedical research.
            </p>
            <ul>
              <li>
                Press releases:
                 
                <a
                  class="primary link"
                  href="http://www.bbc.co.uk/programmes/b08vwn6w"
                >
                  BBC Radio 4 - "inside Science"
                </a>
                 | 
                <a
                  class="primary link"
                  href="https://www.newscientist.com/article/2138671-research-on-male-animals-prevents-women-from-getting-best-drugs/"
                >
                  New Scientist
                </a>
                 | 
                <a
                  class="primary link"
                  href="http://www.dailymail.co.uk/wires/reuters/article-4640316/Sexual-equality-medical-research-long-overdue-study-finds.html"
                >
                  Daily Mail
                </a>
                 | 
                <a
                  class="primary link"
                  href="https://www.sciencedaily.com/releases/2017/06/170626124505.htm"
                >
                  Science Daily
                </a>
                 | 
                <a
                  class="primary link"
                  href="http://fortune.com/2017/06/26/sexual-bias-medical-research/"
                >
                  Fortune
                </a>
                 | 
                <a
                  class="primary link"
                  href="//www.mousephenotype.org/news/study-of-unprecedented-size-reveals-how-sex-blindspot-could-misdirect-medical-research/"
                >
                  IMPC
                </a>
                 | 
                <a
                  class="primary link"
                  href="http://www.ebi.ac.uk/about/news/press-releases/sexual-dimorphism-dilemma"
                >
                  EMBL-EBI
                </a>
                 | 
                <a
                  class="primary link"
                  href="http://www.sanger.ac.uk/news/view/study-reveals-how-sex-blindspot-could-misdirect-medical-research"
                >
                  Sanger
                </a>
              </li>
              <li>
                <a
                  class="primary link"
                  href="https://zenodo.org/record/260398#.WVIkChPys-c"
                >
                  Supporting material to enable replicable analysis
                </a>
              </li>
              <li>
                <a
                  class="primary link"
                  href="https://www.nature.com/articles/ncomms15475"
                >
                  Publication: Nature Communications 26 June 2017
                </a>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
    <div
      class="card"
    >
      <h1>
        <strong>
          Sexual dimorphism in 14,250 wildtype mice
        </strong>
      </h1>
      <div
        class="row"
      >
        <div
          class="col-6"
        >
          <div
            style="position: relative; width: 100%; height: 300px;"
          >
            <canvas
              height="0"
              role="img"
              style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
              width="0"
            />
          </div>
        </div>
        <div
          class="col-6"
        >
          <div
            style="position: relative; width: 100%; height: 300px;"
          >
            <canvas
              height="0"
              role="img"
              style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
              width="0"
            />
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col"
        >
          <p>
            The impact of sex on phenotypes of wild-type C57BL/6N mice measured using the IMPC adult phenotype pipeline at ten global research centers.
          </p>
          <p>
            Sex effect (5% FDR) was detected in:
          </p>
          <ul>
            <li>
              9.9% categorical phenotypes (545 examined, e.g. abnormal corneal opacity)
            </li>
            <li>
              56.6% continuous phenotypes (903 examined, e.g. blood glucose levels)
            </li>
          </ul>
        </div>
      </div>
    </div>
    <div
      class="card"
    >
      <h1>
        <strong>
          Sexual dimorphism in 2,186 knockout mouse strains
        </strong>
      </h1>
      <div
        class="row"
      >
        <div
          class="col-6"
        >
          <div
            style="position: relative; width: 100%; height: 300px;"
          >
            <canvas
              height="0"
              role="img"
              style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
              width="0"
            />
          </div>
        </div>
        <div
          class="col-6"
        >
          <div
            style="position: relative; width: 100%; height: 300px;"
          >
            <canvas
              height="0"
              role="img"
              style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
              width="0"
            />
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col"
        >
          <p>
            The impact of sex on phenotypes of knockout C57BL/6N mouse lines measured using the IMPC adult phenotype pipeline at ten global research centers. Only strains with an abnormal phenotype attributable to the knockout allele (i.e genotype effect) were included.
          </p>
          <p>
            Sex modified the genotype effect on mutant phenotypes in:
          </p>
          <ul>
            <li>
              13.3% categorical phenotypes (1,220 examined; 20% FDR)
            </li>
            <li>
              17.7% continuous phenotypes (7,929 examined; 5% FDR)
            </li>
          </ul>
        </div>
      </div>
    </div>
    <div
      class="card"
    >
      <h1>
        <strong>
          Approach
        </strong>
      </h1>
      <p>
        In 2017, the IMPC analysed phenotype data collected over a 5 year period from 14,250 wildtype and 40,192 mutant mice representing 2,186 knockout lines.
      </p>
      <ul>
        <li>
          Weight was included as a covariate in the continuous data set analyses, as body size is dimorphic between male and female mice and many continuous traits correlate with body weight.
        </li>
        <li>
          For wild-type mice, males and females within a center were compared.
        </li>
        <li>
          For the mutant mice, lines were selected (1) if there was an abnormal phenotype associated with the knockout allele (2) if significant, whether sex influenced the phenotype.
        </li>
      </ul>
      <p>
        The wild-type analysis is from the original manuscript. The mutant line analysis is updated with each IMPC data release.
      </p>
    </div>
    <div
      class="card"
    >
      <h1>
        <strong>
          Vignettes
        </strong>
      </h1>
      <div
        class="row"
      >
        <div
          class="col"
        >
          <div
            style="text-align: center;"
          >
            <h3>
              HDL Cholesterol
            </h3>
            <span>
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-arrow-up-long"
                data-icon="arrow-up-long"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 384 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M214.6 9.4c-12.5-12.5-32.8-12.5-45.3 0l-128 128c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L160 109.3V480c0 17.7 14.3 32 32 32s32-14.3 32-32V109.3l73.4 73.4c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3l-128-128z"
                  fill="currentColor"
                />
              </svg>
               in
               
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-mars"
                data-icon="mars"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M289.8 46.8c3.7-9 12.5-14.8 22.2-14.8H424c13.3 0 24 10.7 24 24V168c0 9.7-5.8 18.5-14.8 22.2s-19.3 1.7-26.2-5.2l-33.4-33.4L321 204.2c19.5 28.4 31 62.7 31 99.8c0 97.2-78.8 176-176 176S0 401.2 0 304s78.8-176 176-176c37 0 71.4 11.4 99.8 31l52.6-52.6L295 73c-6.9-6.9-8.9-17.2-5.2-26.2zM400 80l0 0h0v0zM176 416a112 112 0 1 0 0-224 112 112 0 1 0 0 224z"
                  fill="currentColor"
                />
              </svg>
              , no effect in
               
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-venus"
                data-icon="venus"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 384 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M80 176a112 112 0 1 1 224 0A112 112 0 1 1 80 176zM224 349.1c81.9-15 144-86.8 144-173.1C368 78.8 289.2 0 192 0S16 78.8 16 176c0 86.3 62.1 158.1 144 173.1V384H128c-17.7 0-32 14.3-32 32s14.3 32 32 32h32v32c0 17.7 14.3 32 32 32s32-14.3 32-32V448h32c17.7 0 32-14.3 32-32s-14.3-32-32-32H224V349.1z"
                  fill="currentColor"
                />
              </svg>
            </span>
          </div>
          <div
            style="position: relative; width: 100%; height: 300px;"
          >
            <canvas
              height="0"
              role="img"
              style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
              width="0"
            />
          </div>
          <div
            class="mt-3"
            style="text-align: center;"
          >
            <a
              class="primary link"
              href="/genes/MGI:1922246"
            >
              <i>
                Usp47
                <sup>
                  tm1b(EUCOMM)Wtsi
                </sup>
              </i>
            </a>
          </div>
        </div>
        <div
          class="col"
        >
          <div
            style="text-align: center;"
          >
            <h3>
              Bone Mineral Density
            </h3>
            <span>
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-arrow-down-long"
                data-icon="arrow-down-long"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 384 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M169.4 502.6c12.5 12.5 32.8 12.5 45.3 0l128-128c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 402.7 224 32c0-17.7-14.3-32-32-32s-32 14.3-32 32l0 370.7L86.6 329.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l128 128z"
                  fill="currentColor"
                />
              </svg>
               in
               
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-mars"
                data-icon="mars"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M289.8 46.8c3.7-9 12.5-14.8 22.2-14.8H424c13.3 0 24 10.7 24 24V168c0 9.7-5.8 18.5-14.8 22.2s-19.3 1.7-26.2-5.2l-33.4-33.4L321 204.2c19.5 28.4 31 62.7 31 99.8c0 97.2-78.8 176-176 176S0 401.2 0 304s78.8-176 176-176c37 0 71.4 11.4 99.8 31l52.6-52.6L295 73c-6.9-6.9-8.9-17.2-5.2-26.2zM400 80l0 0h0v0zM176 416a112 112 0 1 0 0-224 112 112 0 1 0 0 224z"
                  fill="currentColor"
                />
              </svg>
              , no effect in
               
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-venus"
                data-icon="venus"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 384 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M80 176a112 112 0 1 1 224 0A112 112 0 1 1 80 176zM224 349.1c81.9-15 144-86.8 144-173.1C368 78.8 289.2 0 192 0S16 78.8 16 176c0 86.3 62.1 158.1 144 173.1V384H128c-17.7 0-32 14.3-32 32s14.3 32 32 32h32v32c0 17.7 14.3 32 32 32s32-14.3 32-32V448h32c17.7 0 32-14.3 32-32s-14.3-32-32-32H224V349.1z"
                  fill="currentColor"
                />
              </svg>
            </span>
          </div>
          <div
            style="position: relative; width: 100%; height: 300px;"
          >
            <canvas
              height="0"
              role="img"
              style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
              width="0"
            />
          </div>
          <div
            class="mt-3"
            style="text-align: center;"
          >
            <a
              class="primary link"
              href="/genes/MGI:1890081"
            >
              <i>
                Foxo3
                <sup>
                  tm1.1(KOMP)Vlcg
                </sup>
              </i>
            </a>
          </div>
        </div>
        <div
          class="col"
        >
          <div
            style="text-align: center;"
          >
            <h3>
              Fructosamine
            </h3>
            <span>
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-arrow-down-long"
                data-icon="arrow-down-long"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 384 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M169.4 502.6c12.5 12.5 32.8 12.5 45.3 0l128-128c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 402.7 224 32c0-17.7-14.3-32-32-32s-32 14.3-32 32l0 370.7L86.6 329.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l128 128z"
                  fill="currentColor"
                />
              </svg>
               in
               
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-mars"
                data-icon="mars"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M289.8 46.8c3.7-9 12.5-14.8 22.2-14.8H424c13.3 0 24 10.7 24 24V168c0 9.7-5.8 18.5-14.8 22.2s-19.3 1.7-26.2-5.2l-33.4-33.4L321 204.2c19.5 28.4 31 62.7 31 99.8c0 97.2-78.8 176-176 176S0 401.2 0 304s78.8-176 176-176c37 0 71.4 11.4 99.8 31l52.6-52.6L295 73c-6.9-6.9-8.9-17.2-5.2-26.2zM400 80l0 0h0v0zM176 416a112 112 0 1 0 0-224 112 112 0 1 0 0 224z"
                  fill="currentColor"
                />
              </svg>
              , 
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-arrow-up-long"
                data-icon="arrow-up-long"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 384 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M214.6 9.4c-12.5-12.5-32.8-12.5-45.3 0l-128 128c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L160 109.3V480c0 17.7 14.3 32 32 32s32-14.3 32-32V109.3l73.4 73.4c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3l-128-128z"
                  fill="currentColor"
                />
              </svg>
               in
               
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-venus"
                data-icon="venus"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 384 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M80 176a112 112 0 1 1 224 0A112 112 0 1 1 80 176zM224 349.1c81.9-15 144-86.8 144-173.1C368 78.8 289.2 0 192 0S16 78.8 16 176c0 86.3 62.1 158.1 144 173.1V384H128c-17.7 0-32 14.3-32 32s14.3 32 32 32h32v32c0 17.7 14.3 32 32 32s32-14.3 32-32V448h32c17.7 0 32-14.3 32-32s-14.3-32-32-32H224V349.1z"
                  fill="currentColor"
                />
              </svg>
            </span>
          </div>
          <div
            style="position: relative; width: 100%; height: 300px;"
          >
            <canvas
              height="0"
              role="img"
              style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
              width="0"
            />
          </div>
          <div
            class="mt-3"
            style="text-align: center;"
          >
            <a
              class="primary link"
              href="/genes/MGI:2446239"
            >
              <i>
                Galnt18
                <sup>
                  tm1b(KOMP)Wtsi
                </sup>
              </i>
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
